test/common/memory/memory_release_test.cc

```diff
@@ -0,0 +1,139 @@
+
+  Envoy::Stats::TestUtil::TestStore stats_;
+  Event::SimulatedTimeSystem time_system_;
+  Api::ApiPtr api_;
+  Event::DispatcherImpl dispatcher_;
+  Envoy::Stats::TestUtil::TestScope scope_;
+  std::unique_ptr<Memory::AllocatorManager> allocator_manager_;
+};
+
+TEST_F(MemoryReleaseTest, ReleaseRateAboveZeroDefaultIntervalMemoryReleased) {
+  size_t initial_allocated_bytes = Stats::totalCurrentlyAllocated();
+  auto a = std::make_unique<unsigned char[]>(MB);
+  auto b = std::make_unique<unsigned char[]>(MB);
+  if (Stats::totalCurrentlyAllocated() - initial_allocated_bytes <= 0) {
+    GTEST_SKIP() << "Skipping test, cannot measure memory usage precisely on this platform.";
+  }
+  auto initial_unmapped_bytes = Stats::totalPageHeapUnmapped();
+  EXPECT_LOG_CONTAINS(
+      "info",
+      "Configured tcmalloc with background release rate: 1048576 bytes per 1000 milliseconds",
+      initialiseAllocatorManager(MB /*bytes per second*/, 0));
+  EXPECT_EQ(MB, AllocatorManagerPeer::bytesToRelease(*allocator_manager_));
+  EXPECT_EQ(std::chrono::milliseconds(1000),
+            AllocatorManagerPeer::memoryReleaseInterval(*allocator_manager_));
+  a.reset();
+  // Release interval was configured to default value (1 second).
+  step(std::chrono::milliseconds(1000));
+  TestUtility::waitForCounterEq(stats_, "memory_release_test.tcmalloc.released_by_timer", 1UL,
+                                time_system_);
+  auto released_bytes_before_next_run = Stats::totalPageHeapUnmapped();
+  b.reset();
+  step(std::chrono::milliseconds(1000));
+  TestUtility::waitForCounterEq(stats_, "memory_release_test.tcmalloc.released_by_timer", 2UL,
+                                time_system_);
+  auto final_released_bytes = Stats::totalPageHeapUnmapped();
+
+#if defined(TCMALLOC) || defined(GPERFTOOLS_TCMALLOC)
+  EXPECT_LT(released_bytes_before_next_run, final_released_bytes);
+  EXPECT_LT(initial_unmapped_bytes, final_released_bytes);
+#else
+  EXPECT_LE(released_bytes_before_next_run, final_released_bytes);
+  EXPECT_LE(initial_unmapped_bytes, final_released_bytes);
+#endif
+}
+
+TEST_F(MemoryReleaseTest, ReleaseRateZeroNoRelease) {
+  auto a = std::make_unique<unsigned char[]>(MB);
+  EXPECT_LOG_NOT_CONTAINS(
+      "info", "Configured tcmalloc with background release rate: 0 bytes 1000 milliseconds",
+      initialiseAllocatorManager(0 /*bytes per second*/, 0));
+  a.reset();
+  // Release interval was configured to default value (1 second).
+  step(std::chrono::milliseconds(3000));
+  EXPECT_EQ(0UL, stats_.counter("memory_release_test.tcmalloc.released_by_timer").value());
+}
+
+TEST_F(MemoryReleaseTest, ReleaseRateAboveZeroCustomIntervalMemoryReleased) {
+  size_t initial_allocated_bytes = Stats::totalCurrentlyAllocated();
+  auto a = std::make_unique<uint32_t[]>(512 * MB);
+  auto b = std::make_unique<uint32_t[]>(512 * MB);
+  if (Stats::totalCurrentlyAllocated() - initial_allocated_bytes <= 0) {
```

Comment:
If I read the code correctly, the two values are unsigned, so will this only fail if the difference is equal to 0?
Consider replacing with `if (Stats::totalCurrentlyAllocated() <= initial_allocated_bytes)`.
Can there be another form of checking whether precise memory measurement is working on the platform?

Reply:
Good catch, fixed. Am not aware of any other simplistic methods to check if test can measure memory allocation for a given platfrom. But since we are testing tcmalloc allocator specifically, i think this check is good enough (e.g. making sure tcmalloc or gperftools_tcmalloc is available and memory usage stats has increased after allocation).